### PR TITLE
Add option to output logs in JSON format

### PR DESCRIPTION
**Pull request by wizardfrag**
## About the changes

This change introduces an option to make logs output in JSON format. This allows automated observability platforms to parse the logs from stdout.

We currently use log shipping from kubernetes to datadog, and without this change, it currently spans multiple log lines for each message output. (See the below screenshots)

Before:
before

After:
after

### Important files

The bulk of the change is in `src/logger.ts`. This change just adds a new class called `JsonLogger` and also adds configuration options and an environment variable

## Discussion points

Happy to discuss anything further. Also if you require any changes or want to redo anything!

### CODE REVIEW

**Comment by sighphyre** [comment]
Hey @wizardfrag, the proxy config options already allow you to take a custom logger implementation. Is there a reason that didn't work for you folks?

**Comment by wizardfrag** [comment]
Hi @sighphyre, you're right, the proxy config options allow you to use a custom logger implementation, however this isn't possible when using the [prebuilt docker image](https://hub.docker.com/r/unleashorg/unleash-proxy). This change allows users of that docker image to use JSON logging output.

I'm happy if you don't want to merge this change though

**Comment by sighphyre** [comment]
@wizardfrag Ah. Right. That's a huge pain! I can see why you'd want this. I'm not a no to making the behaviour better here but what do you think about instead of swapping this out for a new log implementation, to rather extend the behaviour of the SimpleLogger class to squash the contents to JSON if that property is set? I think that'd avoid the potential issue for others where you turn the new env var on but provide a log implementation and get very, very confused as to why your logger is not working.

**Comment by wizardfrag** [comment]
> extend the behaviour of the SimpleLogger class to squash the contents to JSON if that property is set?

That does sound like a good idea.

> I think that'd avoid the potential issue for others where you turn the new env var on but provide a log implementation and get very, very confused as to why your logger is not working.

The way I implemented the config at the moment is that the `logger` option overrides the JSON config, so it would work like before, unless the logger isn't provided. Although this would just invert what you said, and people would wonder why the JSON logger isn't working!

I'd be happy to change it so the SimpleLogger just changes how it logs, though...

The other alternative is that we build our own docker image around this library, and pass our own logger, which also works

**Comment by sighphyre** [comment]
> Although this would just invert what you said, and people would wonder why the JSON logger isn't working!

Also fair! We can probably just put that in the description. I don't think it's _that_ big of a deal.

I think your arguments for the change are good, if this is better for your team than making your own container and if you're willing to run with this and make the changes to SimpleLogger instead I would be most happy to accept this change.

**Comment by wizardfrag** [comment]
Hi, sorry for the delay in reply, I've changed it so that SimpleLogger takes an argument and is used instead of a new logger. Please let me know if anything needs changing, etc!

File: README.md [hunk 1]
```diff
@@ -118,6 +118,7 @@ Keep-Alive: timeout=5
...
 | logger                    | n/a                              | SimpleLogger    |    no    | Register a custom logger.                                                                                                                                                                                                                                                                                                                                                                                                                                                |
+| useJsonLogger             | `JSON_LOGGER`                    | false    |    no    | Use the one-line JSON logger.                                                                                                                                                                                                                                                                                                                                                                                                                                                |
```
Comment (sighphyre) [inline README.md:121]: I think this reflects the new behavior better, WDYT?
```suggestion
| useJsonLogger             | `JSON_LOGGER`                    | false           |    no    | Sets the default logger to log as one-line JSON. This has no effect if a custom logger is provided.                                                                                                                                                                                                                                                                                                                                                                                                                                               |
```

**Comment by sighphyre** [comment]
Hey no worries, completely understand. I have left one small suggestion on the description of the new behaviour, feel free to disagree but if we get that done I think this looks great, happy to accept it

File: README.md (hunk 1 above)
Comment (wizardfrag) [reply README.md:121]: Yep, that makes a lot more sense!

**Pull request merged**